src/transformStateWithClones.js

```diff
@@ -1,13 +1,57 @@
 'use strict';
 
-/**
+/*
  * @param {Object} state
  * @param {Object[]} actions
  *
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
-}
+  const result = [];
+
+  for (const action of actions) {
+    const { type, extraData, keysToRemove } = action;
+    const typeAdd = 'addProperties';
+    const typeRemove = 'removeProperties';
+    const typeClear = 'clear';
+    const newState = result.length === 0
+      ? { ...state }
+      : { ...(result[result.length - 1]) };
```

Comment:
we can just create a `cloneState` outside the loop `const cloneState = {...state}`